optaplanner-core/src/main/java/org/optaplanner/core/impl/io/jaxb/GenericJaxbIO.java

```diff
@@ -0,0 +1,318 @@
+        this(rootClass, DEFAULT_INDENTATION);
+    }
+
+    public GenericJaxbIO(Class<T> rootClass, int indentation) {
+        Objects.requireNonNull(rootClass);
+        this.rootClass = rootClass;
+        this.indentation = indentation;
+        try {
+            jaxbContext = JAXBContext.newInstance(rootClass);
+            marshaller = jaxbContext.createMarshaller();
+            marshaller.setProperty(Marshaller.JAXB_FORMATTED_OUTPUT, true);
+            marshaller.setProperty(Marshaller.JAXB_ENCODING, StandardCharsets.UTF_8.toString());
+        } catch (JAXBException jaxbException) {
+            String errMessage = String.format("Unable to create JAXB Marshaller for a root element class (%s).",
+                    rootClass.getName());
+            throw new OptaPlannerXmlSerializationException(errMessage, jaxbException);
+        }
+    }
+
+    @Override
+    public T read(Reader reader) {
+        Objects.requireNonNull(reader);
+        try {
+            return (T) createUnmarshaller().unmarshal(reader);
+        } catch (JAXBException jaxbException) {
+            String errMessage = String.format(ERR_MSG_READ, rootClass.getName());
+            throw new OptaPlannerXmlSerializationException(errMessage, jaxbException);
+        }
+    }
+
+    public T readAndValidate(Reader reader, String schemaResource) {
+        Objects.requireNonNull(reader);
+        String nonNullSchemaResource = Objects.requireNonNull(schemaResource);
+        SchemaFactory schemaFactory = SchemaFactory.newInstance(XMLConstants.W3C_XML_SCHEMA_NS_URI);
+        Schema schema;
+        try {
+            schema = schemaFactory.newSchema(GenericJaxbIO.class.getResource(nonNullSchemaResource));
+        } catch (SAXException e) {
+            throw new IllegalArgumentException("Unable to read input schema resource (" + nonNullSchemaResource + ")", e);
+        }
+
+        Unmarshaller unmarshaller = createUnmarshaller();
+        unmarshaller.setSchema(schema);
+        ValidationEventStringCollector validationEventHandler = new ValidationEventStringCollector();
+
+        try {
+            unmarshaller.setEventHandler(validationEventHandler);
+        } catch (JAXBException jaxbException) {
+            String errMessage = String.format("Unable to set validation event handler to the unmarshaller for "
+                    + "a root element class (%s).", rootClass.getName());
+            throw new OptaPlannerXmlSerializationException(errMessage, jaxbException);
+        }
+
+        try {
+            return (T) unmarshaller.unmarshal(reader);
+        } catch (JAXBException jaxbException) {
+            String errMessage = String.format(ERR_MSG_READ, rootClass.getName());
+            if (validationEventHandler.hasEvents()) {
+                String errMessageWithValidationEvents = errMessage + "\n" + validationEventHandler.reportAll();
+                throw new OptaPlannerXmlSerializationException(errMessageWithValidationEvents, jaxbException);
```

Comment:
Read didn't fail here. Validation failed. The error message should reflect that.